### PR TITLE
Hardcode custom headers

### PR DESCRIPTION
**Pull request by keur**
Resolves https://github.com/brave/brave-browser/issues/16455

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [x] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (simonhong) [inline components/brave_referrals/browser/brave_referrals_service.cc:564]: nit: How about renaming to `SetReferralHeaders()` or something like as it doesn't fetch at all from network?

**Review by simonhong — APPROVED** [review]
++ with one nit. I can't find more commnet
This is rubber stamp to prevent blocking merge after getting sufficient approve from others :)

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc]: fixed

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (bsclifton) [inline components/brave_referrals/browser/brave_referrals_service.cc:75]: Would defer to @fmarier on this (is this comment accurate (no privacy risk))

**Review by bsclifton — APPROVED** [review]
Left a note for @fmarier but LGTM! ++

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (fmarier) [reply components/brave_referrals/browser/brave_referrals_service.cc:75]: This comment is a great idea. This feature has been misunderstood ever since it came out I think.

My suggestion would be to change this sentence:
>  Since there are tens of millions of Brave users, this presents no privacy risk

to this:
> All Brave users send the exact same header and so this cannot be used for tracking individual users. See https://github.com/brave/brave-browser/wiki/Custom-Headers for more information.

That way we simply stick to the facts, which speak for themselves IMHO.

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (fmarier) [inline components/brave_referrals/browser/brave_referrals_service.cc:71]: I'd say "This header". There are multiple values (possibly for no good reason, see https://github.com/brave/brave-browser/issues/16456), but it's technically the same header.

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (fmarier) [inline components/brave_referrals/browser/brave_referrals_service.cc:115]: Is there an easy way we can call this function on startup to clean up the now-unnecessary file that's on disk for existing clients? (for example, in `BraveReferralsService::Start()`)

**Review by fmarier — CHANGES_REQUESTED** [review]
Thanks for picking this so promptly up @keur !

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (fmarier) [inline components/brave_referrals/browser/brave_referrals_service.cc:98]: `cookieNames` is empty for all three partners. It would be nice to try and remove this dead code since we're not using it and some people might be wondering why it looks like we're doing something with cookies.

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc:115]: Let’s keep this PR scoped the hard-coding the headers. This is unrelated.

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc]: fixed

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc]: fixed

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:612]: I guess we can erase the loader as well

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:599]: we can drop the constant `kBraveReferralsHeadersPath` as well

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:82]: these are prohibited by the style guide, see https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

pls change to `std::array` or plain C arrays

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:571]: can't we just hardcode the whole value as a string and then parse the `base::Value` from json in one line?

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:106]: i'd name it expiration_ms

**Comment by iefremov** [comment]
After looking at the code I've found out that these headers are only used in `OnBeforeStartTransaction_ReferralsWork`, so we can erase even more stuff: `BraveRequestHandler::OnReferralHeadersChanged`, the field in `brave::BraveRequestInfo`, the preference. Just check the hardcoded list in `OnBeforeStartTransaction_ReferralsWork`

**Review by iefremov — CHANGES_REQUESTED** [review]
It seems we can drop even more legacy code

**Review by fmarier — APPROVED** [review]

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc:612]: fixed

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc]: fixed

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc:565]: this is a style thing, and i prefer what i have. parsing JSON not covered by tests. at least with this we get compile-time type checking

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc]: fixed

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc:599]: fixed

File: browser/net/brave_referrals_network_delegate_helper.cc
Comment (iefremov) [inline browser/net/brave_referrals_network_delegate_helper.cc:28]: I was trying to explain that this is preference is also obsolete now. Since the list is permanent, there is no need to save the hardcoded value to prefs and then load, you can just filter URLs and add hardcoded headers in this function. Or do it in `BraveReferralsService::GetMatchingReferralHeaders`, but without juggling with the pref (that could actually be removed from both local state and profile prefs prefs now)

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [reply components/brave_referrals/browser/brave_referrals_service.cc:565]: so we can remove fake header data from the test and do real checks for real hardcoded domains. This will increase safety and reduce boilerplate.

File: browser/net/brave_referrals_network_delegate_helper.cc
Comment (keur) [reply browser/net/brave_referrals_network_delegate_helper.cc]: i made it a singleton. fixed

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:697]: oh, this is interesting. Do we serve any headers from `/promo/initialize/nonua"`? @fmarier

File: components/brave_referrals/browser/brave_referrals_service.h
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.h:40]: there is really no need to write the singleton boilerplate. Just a function that internally uses a static vector (or even better, flat_map) with hardcoded values would be enough.

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:85]: is it used at all?

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:152]: personally, I would make global
```
constexpr auto kHeaders = base::MakeFixedFlatMap<base::StringPiece, base::StringPiece> {
 {"eaff.com", "eaff}, {"stg.eaff.com", "eaff"} , {"uphold.com", "uphold}, ...
}
```
then I would iterate over keys and match the domain and then get the header value in one line (if this is everything we need).

**Review by iefremov — APPROVED** [review]

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc:697]: good catch. no we don't (i authored the current version of laptop-updates). going to remove this too

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc:83]: yes

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc]: fixed

Comment (keur) [reply components/brave_referrals/browser/brave_referrals_service.cc]: confirmed with @aekeus this was good to remove

File: components/brave_referrals/browser/brave_referrals_service.cc
Comment (iefremov) [inline components/brave_referrals/browser/brave_referrals_service.cc:746]: that means we don't need `GetMatchingReferralHeaders` overloads?

**Pull request merged**
